core/src/main/java/cucumber/runtime/Runtime.java

```diff
@@ -40,39 +53,139 @@ public Runtime(Plugins plugins,
         this.runtimeOptions = runtimeOptions;
         this.filters = filters;
         this.bus = bus;
-        this.runner = runnerSupplier.get();
+        this.runnerSupplier = runnerSupplier;
         this.featureSupplier = featureSupplier;
         exitStatus.setEventPublisher(bus);
     }
 
     public void run() {
         List<CucumberFeature> features = featureSupplier.get();
         bus.send(new TestRunStarted(bus.getTime()));
-        for (CucumberFeature feature : features) {
-            feature.sendTestSourceRead(bus);
-        }
-
+        
         StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
-
-        runner.reportStepDefinitions(stepDefinitionReporter);
-
-        for (CucumberFeature cucumberFeature : features) {
-            runFeature(cucumberFeature);
+        runnerSupplier.get().reportStepDefinitions(stepDefinitionReporter);
+        
+        final int requestedThreads = runtimeOptions.getThreads();
+        if (!features.isEmpty()) {
+            for (CucumberFeature feature : features) {
+                feature.sendTestSourceRead(bus);
+            }
+            ParallelFeatureRunner.run(new FeatureRunner(filters, runnerSupplier), features, requestedThreads);
```

Comment:
Can you explain why you are executing features in parallel and not pickles? This adds allot of complexity which could be naively be done by:

```java
        for (CucumberFeature feature : features) {
            for (final PickleEvent pickleEvent : compiler.compileFeature(feature)) {
                if (filters.matchesFilters(pickleEvent)) {
                    executor.submit(new Runnable() {
                        @Override
                        public void run() {
                            runnerSupplier.get().runPickle(pickleEvent);
                        }
                    });
                }
            }
        }
```

(this doesn't await completion, but that is easy to fix)

Reply:
I wanted to create a callable to do the work e.g. pulling the test feature to be run from the queue.

Didn’t think about breaking it down into pickle level (individual scenario right?) but could be done I suppose as it increases the benefits if a feature has many scenarios but another only has 1

Waiting for the result could be done by submitting a Future instead

Again I was just trying to keep each class doing one thing, hence extracting the classes and working with the features as in parallel as opposed to pickles

I’ll have a play with this and come back

Reply:
You may notice that the formatters will break because they expect the `TestCase*` events from a the scenarios in a feature to be in order.

I think we can add another event bus for that, one which sorts the sets of TestCaseEvents by their Uri's.

Reply:
> Waiting for the result could be done by submitting a Future instead

If the nested for loops adds the Runnables to a list instead  you could call `executor.invokeAll(tasks)`.

Reply:
Have made the changes so that `PickleEvents` are now run in parallel.

The `Formatter`s, well their tests, appear to be OK. Though they're not running in parallel.  When I had done this before the tests didn't always pass in Travis.

The events are still being sent in the correct order and synchronized accordingly
- `TestCaseStarted`
- `TestStepStarted`
- `EmbedEvent`
- `WriteEvent`
- `TestStepFinished`
- `TestCaseFinished`